presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -46,7 +47,16 @@
     @Override
     public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
     {
-        WindowNode child = captures.get(CHILD);
+        // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+        // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+        // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+        // input symbols.
```

Comment:
> What about other nodes which could be skipped during WindowNode merging, like DistintNode, FilterNode?

That would be nice to handle too, however we don't see anything like that in production, so it does not seem the most important problem to solve for now.
For `FilterNode`, actually, this should be handled by other rules, i.e. pushdown.

```
Also what about sorting when WindowNodes are separated with ProjectNode, like below?

sum over (partition by A order by C)
project()
avg over (partition by B)
project()
sum over (partition by A order by D)
```

I believe what would happen then is to push down one of the projections below the WN, let's say to this intermediate stage:
```
sum over (partition by A order by C)
avg over (partition by B)
project()
project()
sum over (partition by A order by D)
```
Then, if `SwapAdjacentWindows` comes to action now it may swap first two windows but it does not matter.
Next iteration of pulling WFs above projections will pull the last WN above two projections to the form of:
```
sum over (partition by A order by C)
avg over (partition by B)
sum over (partition by A order by D)
project()
project()
```
Then merging/swapping can continue until we get:
```
avg over (partition by B)
sum over (partition by A order by D)
sum over (partition by A order by C)
project()
project()
```